### PR TITLE
Token assigned feature

### LINKED ISSUE #5: Subscribe and process`TokensAssigned` Event
The `TokensAssigned` event would be emitted by the Wallet API service. The format of this event is to be finalized and not yet concrete. But the idea is, captures processed and verified would have eligible tokens and these tokens are owned by or associated with wallets. Thus, capture records are associated to the wallets and tokens so that clients using the data can query capture features by wallet or token. 

When a creation or transfer of tokens occurs in wallet api, TokensAssigned event is generated. The format of this event will be roughly of the following kind (concrete format will be known soon but will be more or less the same as follows):

{
  "type": "TokensAssigned",
  "wallet_name": "joeswallet",
  "entries": [ 
{  "capture_id": "63e00bca-8eb0-11eb-8dcd-0242ac130003", "token_id": "9d7abad8-8eb0-11eb-8dcd-0242ac130003"  },
{ "capture_id": "8533b704-8eb0-11eb-8dcd-0242ac130003", "token_id":"a5799d94-8eb0-11eb-8dcd-0242ac130003" } ]
}

On handling this event, process and update `capture_feature` with the wallet name and that of the token_id. When the processing is complete, mark the event as `handled`.

### PR DESCRIPTION
**Pull request by dadiorchen**
resolve #5 

@arunbakt I think I still didn't fully understand your idea in terms of the function style coding, so there might be some wrong implementation in the PR, let me know! 

I added an integration test, and run the test locally to test the code.

### CODE REVIEW

File: __tests__/integration/tokenAssigned.spec.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    const wallet_name_new = "newone";
+    const message = {
+      "type": "TokensAssigned",
+      "wallet_name": wallet_name_new,
+      "entries": [
+        { "capture_id": capture_id, "token_id": token_id },
+      ],
+    }
+    publish("token-assigned", undefined, message, (e) => console.log("result:", e));
+    await new Promise(r => setTimeout(() => r(), 2000));
```
Comment (dadiorchen) [inline __tests__/integration/tokenAssigned.spec.js:47]: This way that sleeping for a fixed time to wait for the message receiving is a little ugly. There might be a better way to do so.

File: package.json [hunk 2]
```diff
@@ -18,7 +18,7 @@
     "server-test": "DEBUG=express:* NODE_LOG_LEVEL=debug nodemon server/serverTest.js",
     "server": "nodemon server/server.js",
     "test-seedDB": "NODE_ENV=test mocha -r dotenv/config dotenv_config_path=.env.test --timeout 10000 --require co-mocha './**/*.spec.js'",
-    "test-integration": "NODE_ENV=test mocha -r dotenv/config dotenv_config_path=.env.test --exit --timeout 20000 --require co-mocha './__tests__/supertest.js'",
+    "test-integration": "NODE_ENV=test mocha -r dotenv/config dotenv_config_path=.env.test --exit --timeout 20000 --require co-mocha './__tests__/integration/**/*.spec.js'",
```
Comment (dadiorchen) [inline package.json:21]: Puting all the integration test in the `__tests__/integration`

File: server/infra/messaging/rabbit-mq-messaging.js [hunk 3]
```diff
@@ -1,9 +1,10 @@
 const Broker = require('rascal').BrokerAsPromised
 const config = require('./config').config
+const log = require("loglevel");
 
 const publish = (async (publicationName, routingKey, payload, resultHandler) => {
-    const broker = await Broker.create(config);
     try {
+        const broker = await Broker.create(config);
```
Comment (dadiorchen) [inline server/infra/messaging/rabbit-mq-messaging.js:7]: I found it's possible that `Broker.create` would throw an error, and if I don't wrap the cod by `try/catch` then nothing will be thrown in the output, so it's hard to debug

File: server/models/capture-feature.js [hunk 4]
```diff
@@ -35,4 +36,10 @@ const createCaptureFeature = (captureFeatureRepo) => (async captureFeature => {
     repository.add(captureFeature);
 });
 
-module.exports = { captureFeatureFromMessage, createCaptureFeature };
+const updateCaptureFeature = (captureFeatureRepo) => (async (captureFeatureIds, captureFeatureUpdateObject) => {
+  log.warn("repo:", captureFeatureRepo);
+  // Because here is using a special fn to update db, so didn't use Repository
+  await captureFeatureRepo.batchUpdate(captureFeatureIds, captureFeatureUpdateObject);
+});
```
Comment (dadiorchen) [inline server/models/capture-feature.js:43]: Casually, I don't know why we need to use `new Repository` there.

File: __tests__/integration/tokenAssigned.spec.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    await knex("capture_feature").insert(capture);
+    const wallet_name_new = "newone";
+    const message = {
+      "type": "TokensAssigned",
+      "wallet_name": wallet_name_new,
+      "entries": [
+        { "capture_id": capture_id, "token_id": token_id },
+      ],
+    }
+    publish("token-assigned", undefined, message, (e) => console.log("result:", e));
```
Comment (arunbakt) [inline __tests__/integration/tokenAssigned.spec.js:46]: could we possibly skip involving the queue  and instead use the registered handler for the event to test it? 
`publish` function is meant for dispatching messages from the originating system. `token-assigned` here is a consumed message and should enter the web-map-query-consumer via `subscribe` function.  I am not sure if this test is really exercising/simulating the flow of message into web-map-query-consumer (`publish` sends out messages to a topic to be consumed by another service and not by itself). I recommend to skip the queues so it is simpler.

File: server/infra/messaging/rabbit-mq-messaging.js [hunk 6]
```diff
@@ -31,4 +32,14 @@ const subscribe = (async (subscriptionName, eventHandler) => {
     }
 });
 
-module.exports = { publish, subscribe };
\ No newline at end of file
+const unsubscribeAll = (async () => {
```
Comment (arunbakt) [inline server/infra/messaging/rabbit-mq-messaging.js:35]: Curious, when would we use `unsubscribeAll`?

File: server/models/capture-feature.js (hunk 4 above)
Comment (arunbakt) [reply server/models/capture-feature.js:43]: The idea was to give  models package with functions to have only local dependencies and explicitly stating what signatures are expected when passing an repository implementation. I agree its need is arguable.

File: server/services/event-handlers.js [hunk 7]
```diff
@@ -21,9 +22,11 @@ const createRawCaptureFeatureHandler = (async (message) => {
   executeRawCreateCaptureFeature(newRawCaptureFeature);
 })
 
+
 const registerEventHandlers = () => {
   subscribe("capture-created", createCaptureFeatureHandler);
   subscribe("raw-capture-created", createRawCaptureFeatureHandler);
+  subscribe("token-assigned", tokenAssignedHandler);
```
Comment (arunbakt) [inline server/services/event-handlers.js:29]: In the intergation test, we could choose to invoke `tokenAssginedHandler` directly to test this flow by not involving rabbitmq.

**Review by arunbakt — COMMENTED** [review]
Hi @dadiorchen , 

Overall it looks good. I left couple of suggestions you could consider looking into (mainly about integration tests)

Arun

File: __tests__/integration/tokenAssigned.spec.js (hunk 5 above)
Comment (dadiorchen) [reply __tests__/integration/tokenAssigned.spec.js:46]: Arun the whole test passed, it proved that the queue and DB process works well, so it's useful, the argument point I guess is that if we worth the effort doing this. 
I prefer to link them all together and test them as a whole, it makes me more confident on the code. If we skip it and call the handler directly, it's more like a unit test, the queue workflow is uncovered.
From another angle, skipped the queue workflow, I still need to test it by myself, by operating the queue manually, it's more inefficient IMO, and another bonus of building this test is that I got a test to cover my code in the future.
Maintaining the test is a question, but a higher level integration test is easier to maintain than a lower level of unit tests, I think.

File: server/infra/messaging/rabbit-mq-messaging.js (hunk 6 above)
Comment (dadiorchen) [reply server/infra/messaging/rabbit-mq-messaging.js:35]: I just used it for test workflow :)

**Pull request merged**